# E-Commerce Logs Analysis

## Project Objective

The objective of this project is to analyze the website logs from an e-commerce store to obtain useful insight into customer behaviour.

## Data Source

The dataset was obtained here: https://www.kaggle.com/datasets/kzmontage/e-commerce-website-logs and the data has an Open Database license: https://opendatacommons.org/licenses/odbl/1-0/

In [4]:
-- Specifying database
USE EcommerceLogsDb

Commands completed successfully.

Total execution time: 00:00:00.003

In [5]:
-- preview of purchase logs table
SELECT TOP 3*
FROM logs;

(3 rows affected)

Total execution time: 00:00:00.007

access_date duration_seconds Proto IP Src_IP_type Src_Pt Bytes Accessed_From country membership Languages Sales Returned Returned_Amount 2016-11-01 09:58:00.0000000 2533 TCP 1.10.195.126 EXT_SERVER 8082 20100 Chrome Albania Normal Albanian 261.96 No 0 2016-11-01 09:59:00.0000000 4034 TCP 1.1.217.211 OPENSTACK_NET 56978 20500 Mozilla Firefox Albania Normal Albanian 731.94 No 0 2016-11-01 09:59:00.0000000 1525 TCP 1.115.198.107 EXT_SERVER 8082 90100 Mozilla Firefox Albania Normal Albanian 14.62 No 0

In [13]:
-- Which country had the highest number of sales and highest value of sales (in $)?
-- showing country percentiles for both metrics using window functions
SELECT country, 
    COUNT(*) AS num_sales, 
    PERCENT_RANK() OVER(ORDER BY COUNT(*)) AS num_sales_percentile, 
    SUM(Sales) AS value_sales, 
    PERCENT_RANK() OVER(ORDER BY SUM(Sales)) AS value_sales_percentile
FROM logs
GROUP BY country
ORDER BY value_sales DESC;

(101 rows affected)

Total execution time: 00:00:00.119

country num_sales num_sales_percentile value_sales value_sales_percentile Austria 2674 1 1296463.3414999906 1 Belgium 2604 0.97 1154321.3705999977 0.99 Argentina 2603 0.96 1140238.6927999987 0.98 Australia 2520 0.94 1139576.1649999968 0.97 United States 2231 0.73 1132496.3800000001 0.96 Brazil 2604 0.97 1129919.2000999975 0.95 Norway 2160 0.68 1129793.889999999 0.94 Netherlands 2292 0.88 1106623.810000002 0.93 Spain 2232 0.74 1102578.460000004 0.92 Puerto Rico 2232 0.74 1089656.3800000006 0.91 Canada 2436 0.92 1078310.4730000005 0.9 United Kingdom 2232 0.74 1069649.1000000013 0.89 Bulgaria 2520 0.94 1068255.9588000018 0.88 Republic of Korea 2232 0.74 1063598.8800000008 0.87 Singapore 2232 0.74 1062442.5199999993 0.86 Suriname 2016 0.58 1037655.410000003 0.85 Mexico 2232 0.74 1034845.1700000012 0.84 Turkmenistan 2088 0.63 1032696.5800000039 0.83 Belize 2352 0.91 1009231.1078000019 0.82 Sweden 2148 0.67 1006956.3000000026 0.81 Chile 2604 0.97 998301.8032000033 0.8 Russia 1952 0.55 990272.1340000018 0.79 Thailand 2004 0.56 984841.0600000018 0.78 Portugal 1944 0.52 955279.5800000016 0.77 Saint Lucia 2016 0.58 953605.8500000038 0.76 Mauritius 2292 0.88 951543.6300000029 0.75 Trinidad and Tobago 1944 0.52 940220.2900000017 0.74 Antigua and Barbuda 2268 0.87 937031.6729000005 0.73 Romania 2004 0.56 933676.0400000019 0.72 Armenia 2086 0.62 918236.707500002 0.71 Barbados 2099 0.66 914011.7824000027 0.7 Uruguay 2016 0.58 898636.5600000003 0.69 New Zealand 2088 0.63 897640.3300000007 0.68 Paraguay 1944 0.52 887757.4500000015 0.67 Luxembourg 2232 0.74 869798.3930000014 0.66 Ukraine 2016 0.58 860557.6200000015 0.65 Bahamas 1932 0.5 858137.5198000026 0.64 Malta 2232 0.74 849198.6836000021 0.63 Panama 1800 0.41 827407.4100000006 0.62 Poland 1728 0.35 823874.8300000004 0.61 Colombia 2514 0.93 823458.3767000027 0.6 Albania 1848 0.44 816714.1692000002 0.59 Guatemala 2160 0.68 811037.0116000031 0.58 Bahrain 1764 0.39 800315.182500001 0.57 Belarus 1763 0.38 793122.7560000019 0.56 Saint Vincent and Grenadines 1800 0.41 782619.9400000017 0.55 Greece 2232 0.74 781112.8694000007 0.54 Slovenia 1511 0.25 779910.2300000001 0.53 Slovakia 1584 0.31 774529.1199999993 0.52 Switzerland 1512 0.27 711132.8300000015 0.51 Ecuador 2232 0.74 704434.9145000005 0.5 Uzbekistan 1584 0.31 700289.7400000006 0.49 South Africa 1440 0.24 699497.384000001 0.48 Japan 2233 0.86 692880.2690000008 0.47 Israel 2232 0.74 678394.283000001 0.46 Grenada 1860 0.45 677864.1923000009 0.45 Iceland 2292 0.88 673693.6802000001 0.44 Guyana 1799 0.4 673308.8582000005 0.43 Kuwait 1800 0.41 668267.5795000002 0.42 Kyrgyzstan 1872 0.47 666682.2039000005 0.41 Ireland 2160 0.68 649537.6800000011 0.4 France 2160 0.68 643784.2059999998 0.39 Kazakhstan 1872 0.47 641065.9040000006 0.38 Hungary 1860 0.45 634635.7702 0.37 Costa Rica 2160 0.68 632901.6622000004 0.36 Finland 2088 0.63 630622.2634000012 0.35 Italy 2232 0.74 625664.3300000008 0.34 Seychelles 1296 0.21 620275.4700000003 0.33 Azerbaijan 1344 0.23 610523.8879999998 0.32 Germany 1872 0.47 601330.1426000008 0.31 El Salvador 1728 0.35 593740.2684 0.3 Lithuania 1572 0.29 572731.2999999988 0.29 Serbia 1296 0.21 564452.1499999986 0.28 Qatar 1068 0.17 554514.8799999999 0.27 Latvia 1511 0.25 549710.9939999995 0.26 Aruba 1176 0.19 537197.9709999991 0.25 Philippines 1079 0.18 522248.86000000004 0.24 Czech Republic 1932 0.5 517375.4399999993 0.23 Cuba 1728 0.35 500943.9629999991 0.22 Georgia 1584 0.31 499963.55559999857 0.21 Denmark 1584 0.31 449218.213999999 0.2 Estonia 1512 0.27 424996.7387999989 0.19 Croatia 1572 0.29 416275.25939999963 0.18 Sri Lanka 792 0.13 395918.2999999992 0.17 Montenegro 720 0.11 357511.14999999985 0.16 Jamaica 1224 0.2 347578.4049999999 0.15 Kiribati 792 0.13 323550.16689999966 0.14 Cyprus 1067 0.16 281314.48579999973 0.13 Turkey 504 0.1 280994.1200000001 0.12 Nicaragua 432 0.08 223247.22000000006 0.11 Maldives 720 0.11 217764.10799999998 0.1 United Arab Emirates 432 0.08 186636.71000000022 0.09 Fiji 792 0.13 186410.4449999999

From the table above, the top 4 countries for num\_sales and value\_sales are Austria, Belgium, Argentina and Australia. The number of sales seems highly correlated with value of sales with some exceptions like the United States which ranks in the 73rd percentile for number of sales and in the 96th percentile for value of sales.

Let's look at the average value of sale per country and compare to the worldwide average. The difference between the average sale value for each country and the worldwide average will also be computed.

In [7]:
-- average sale value ($) per country
SELECT  
    country, AVG(Sales) AS avg_sale_value,
    -- subquery in SELECT to determine worldwide avg sale value ($)
    (SELECT AVG(Sales) FROM logs) AS world_avg_sale_value,
    -- difference bewteen country avg and worldwide avg with another SELECT subquery
    (AVG(Sales) - (SELECT AVG(Sales) FROM logs)) AS difference
FROM logs
GROUP BY country
ORDER BY difference DESC;

(101 rows affected)

Total execution time: 00:00:00.093

country avg_sale_value world_avg_sale_value difference Turkey 557.528015873016 411.346448791984 146.18156708103146 Norway 523.0527268518514 411.346448791984 111.70627805986686 Qatar 519.208689138577 411.346448791984 107.8622403465925 Nicaragua 516.7759722222223 411.346448791984 105.4295234302378 Slovenia 516.1550165453339 411.346448791984 104.80856775334934 Suriname 514.7100248015873 411.346448791984 103.36357600960281 United States 507.6182787987448 411.346448791984 96.2718300067603 Russia 507.311544057377 411.346448791984 95.96509526539245 Sri Lanka 499.8968434343436 411.346448791984 88.5503946423591 Montenegro 496.54326388888893 411.346448791984 85.1968150969044 Turkmenistan 494.5864846743293 411.346448791984 83.24003588234478 Spain 493.9867652329743 411.346448791984 82.64031644098975 Thailand 491.4376546906189 411.346448791984 80.09120589863437 Portugal 491.3989609053497 411.346448791984 80.0525121133652 Slovakia 488.97040404040365 411.346448791984 77.62395524841912 Oman 488.639675925926 411.346448791984 77.29322713394146 Puerto Rico 488.19730286738303 411.346448791984 76.8508540753985 South Africa 485.76207222222195 411.346448791984 74.41562343023742 Austria 484.8404418474195 411.346448791984 73.493993055435 Philippines 484.01191844300286 411.346448791984 72.66546965101833 Trinidad and Tobago 483.65241255144014 411.346448791984 72.30596375945561 Netherlands 482.8201614310642 411.346448791984 71.47371263907968 United Kingdom 479.2334677419357 411.346448791984 67.8870189499512 Seychelles 478.60761574074047 411.346448791984 67.26116694875594 Poland 476.77941550925925 411.346448791984 65.43296671727472 Republic of Korea 476.52279569892437 411.346448791984 65.17634690693984 Singapore 476.00471326164853 411.346448791984 64.658264469664 Saint Lucia 473.018774801587 411.346448791984 61.6723260096025 Switzerland 470.32594576719583 411.346448791984 58.979496975211305 Sweden 468.7878491620107 411.346448791984 57.441400370026145 Romania 465.90620758482993 411.346448791984 54.5597587928454 Mexico 463.64030913978485 411.346448791984 52.29386034780032 Panama 459.6707833333328 411.346448791984 48.324334541348264 Aruba 456.80099574829933 411.346448791984 45.4545469563148 Paraguay 456.66535493827075 411.346448791984 45.318906146286224 Azerbaijan 454.2588452380954 411.346448791984 42.91239644611085 San Marino 453.7129166666666 411.346448791984 42.36646787468209 Bahrain 453.69341411564585 411.346448791984 42.34696532366132 Australia 452.21276388888845 411.346448791984 40.86631509690392 Belarus 449.8711038003401 411.346448791984 38.52465500835558 Uruguay 445.7522619047622 411.346448791984 34.40581311277765 Bahamas 444.1705589026912 411.346448791984 32.82411011070667 Belgium 443.28777672811094 411.346448791984 31.941327936126413 Canada 442.6561876026271 411.346448791984 31.309738810642557 Uzbekistan 442.1021085858583 411.346448791984 30.755659793873747 Albania 441.94489675324695 411.346448791984 30.598447961262423 Armenia 440.19017617449674 411.346448791984 28.843727382512213 Argentina 438.0479034959661 411.346448791984 26.701454703981597 Serbia 435.534066358025 411.346448791984 24.187617566040444 Barbados 435.45106355407376 411.346448791984 24.104614762089227 Cabo Verde 435.26276190476193 411.346448791984 23.916313112777402 Saint Vincent and Grenadines 434.7888555555559 411.346448791984 23.442406763571398 Brazil 433.9167435099842 411.346448791984 22.570294717999673 United Arab Emirates 432.02942129629656 411.346448791984 20.682972504312033 New Zealand 429.9043726053638 411.346448791984 18.55792381337926 Belize 429.0948587585031 411.346448791984 17.74840996651858 Ukraine 426.863898809524 411.346448791984 15.517450017539488 Bulgaria 423.91109476190445 411.346448791984 12.564645969919923 Mauritius 415.1586518324606 411.346448791984 3.8122030404760494 Antigua and Barbuda 413.1532949294535 411.346448791984 1.8068461374689946 Kiribati 408.5229380050506 411.346448791984 -2.823510786933923 Saint Kitts and Nevis 404.76092592592585 411.346448791984 -6.

In [17]:
SELECT Sales, COUNT(1) AS value_counts
FROM logs
GROUP BY country, Sales
HAVING country = 'Canada';

(390 rows affected)

Total execution time: 00:00:00.049

Sales value_counts 0 399 2.072 1 2.296 1 2.304 1 2.512 1 2.544 1 2.556 1 2.688 1 2.92 1 3.024 1 3.108 1 3.264 1 3.304 1 3.318 1 3.408 1 3.592 1 3.62 1 3.9 1 4.356 1 4.36 1 4.448 1 4.512 1 4.6 1 4.923 1 4.992 1 5.184 2 5.22 1 5.23 53 5.68 1 5.792 1 6.672 1 6.688 1 7.24 1 7.764 1 7.83 1 7.968 1 8.096 1 8.134 1 8.34 1 8.4 2 8.544 1 8.56 1 8.57 1 8.784 1 9.26 1 9.522 1 9.618 1 10.048 1 10.11 1 10.312 1 10.44 1 10.46 53 10.48 1 10.56 2 10.776 2 11.648 1 11.68 1 11.73 16 12.42 1 12.536 1 12.544 1 12.576 1 12.84 1 12.9 1 12.96 2 13.092 1 13.12 1 13.68 1 13.904 1 14.016 1 14.224 1 14.4 1 14.624 1 14.7 1 14.88 1 14.94 1 15.15 25 15.552 4 15.69 62 15.76 1 15.84 1 16.776 2 16.99 1 17.05 1 17.46 1 17.48 1 17.496 1 17.94 2 18.06 1 18.24 1 18.432 1 18.54 1 18.84 1 18.864 1 19.04 1 19.152 1 19.44 2 19.46 1 19.9 1 20.32 1 20.46 1 20.568 1 20.7 1 20.736 1 20.768 1 20.92 60 21.36 1 21.4 1 22.72 1 23.1 1 23.16 1 23.34 1 23.46 19 24.4 1 26.15 61 26.4 1 26.49 1 27.18 1 27.496 1 28.4 1 28.9 1 29.24 1 29.472 1 29.9 1 30.3 19 30.36 1 30.93 1 33.48 1 33.568 1 33.74 1 33.9 1 33.93 1 34.44 1 35.04 1 35.168 3 35.19 24 35.2 1 35.4 1 35.48 1 35.568 1 35.84 38 36.112 1 36.56 1 36.63 1 37.4 1 37.68 1 38.34 1 38.71 1 38.88 1 39.872 1 39.9 1 39.92 1 40.032 1 40.29 1 40.66 61 41.72 1 41.88 1 42.048 1 43.19 1 43.5 1 43.6 1 43.98 1 44.82 1 44.84 1 45.04 1 45.36 1 45.45 24 45.528 1 45.66 1 45.78 1 45.98 1 46.688 1 46.8 1 46.92 24 47.79 1 48.712 1 48.784 1 49.12 1 51.45 2 51.55 1 52.064 1 52.512 1 53.248 1 53.88 1 54.96 1 55.5 1 55.936 1 56.7 1 57.75 1 58.65 20 58.73 1 59.98 1 60.6 17 61.4 1 62.8 1 62.94 1 63.84 1 64.704 1 66.36 1 67.144 1 67.71 1 67.78 1 67.84 1 67.92 1 68.46 1 68.72 1 68.81 1 69.48 1 69.5 1 70.26 1 71.372 1 71.68 43 72.64 1 75.75 16 75.84 1 77.88 1 79.056 1 79.14 2 79.984 1 79.99 1 81.32 64 81.98 1 84.84 1 85.9 1 85.96 1 86.26 1 86.304 1 88.768 1 90.48 1 90.57 1 91.275 1 93.344 1 95.616 1 96.08 2 99.372 1 99.846 1 99.98 1 101.12 1 105.52 1 105.584 1 105.98 1 107.52 53 107.648 1 107.976 1 109.592 1 110.98 1 113.328 1 114.95 1 117.488 1 119.616 1 119.9 1 121.98 63 123.96 1 124.2 1 128.85 1 129.92 1 136.464 1 137.62 1 139.04 1 143.36 47 144.12 1 147.168 1 158.13 1 162.64 68 164.22 1 166.44 1 167.535 1 167.952 1 170.352 1 173.94 1 179.2 41 182.91 1 186.144 1 186.304 1 186.912 1 190.92 1 191.6 1 195.64 1 196.776 1 199.836 1 199.96 1 199.98 1 200.795 1 203.3 56 207.984 1 209.979 1 210.008 1 210.564 1 212.058 1 212.94 3 213.48 1 214.95 1 221.06 1 222.32 1 231.98 1 232.55 1 239.96 1 239.976 1 242.94 1 249.95 1 254.058 1 258.576 1 283.92 1 287.97 1 299.98 1 302.67 1 314.55 1 328.3992 1 332.94 1 333.576 1 333.999 1 347.58 1 353.568 1 359.499 1 362.94 1 371.168 1 375.34 1 377.97 1 383.976 1 391.98 1 397.6 1 398.4 1 401.59 1 407.976 1 431.976 1 461.97 1 464.292 1 465.18 1 466.158 1 467.04 1 467.46 1 498.26 1 532.3992 1 535.41 1 549.99 1 563.94 1 569.058 1 584.82 1 600.16 133 600.17 44 601.3 1 605.34 1 629.1 1 629.93 1 653.55 1 665.88 1 687.4 1 689.408 1 697.16 1 699.98 1 706.86 1 714.3 1 716 1 717.12 1 720.76 1 744.1 1 748.752 1 772.47 2 773.94 1 791.964 1 842.376 1 897.15 1 900.24 147 911.424 1 955.24 1 1001.584 1 1044.63 1 1050 15 1097.544 1 1214.85 1 1268.82 1 1317.492 1 1325.76 1 1325.85 1 1352.3976 1 1367.84 1 1500.4 135 1502.376 1 1516.2 1 1781.682 1 1800.51 35 1801.632 1 1889.99 1 2100 20 2321.9 1 2400.68 41 2799.96 1 3000.85 40 3930.072 1 5250 27

In [20]:
SELECT COUNT(Sales)
FROM logs
WHERE Sales = 0;

(1 row affected)

Total execution time: 00:00:00.050

(No column name) 41338

In [8]:
-- median sale value ($) per country
SELECT  
    DISTINCT country,
    PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY Sales) OVER(PARTITION BY country) AS median,
    PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY Sales) OVER() AS world_median
FROM logs
ORDER BY median DESC;

(101 rows affected)

Total execution time: 00:00:01.614

country median world_median San Marino 121.98 46.92 Mongolia 94.41999999999999 46.92 Portugal 81.32 46.92 Nicaragua 81.32 46.92 United States 81.32 46.92 South Africa 81.32 46.92 Serbia 81.32 46.92 Aruba 81.32 46.92 Spain 81.32 46.92 Slovakia 81.32 46.92 Norway 81.32 46.92 Suriname 81.32 46.92 Trinidad and Tobago 81.32 46.92 Austria 81.32 46.92 Romania 81.32 46.92 Qatar 81.32 46.92 Slovenia 81.32 46.92 Montenegro 81.32 46.92 Bosnia and Herzegovina 78.535 46.92 Turkmenistan 75.75 46.92 Turkey 75.75 46.92 Mexico 75.75 46.92 Puerto Rico 75.75 46.92 Singapore 75.75 46.92 Bahamas 75.75 46.92 Seychelles 75.75 46.92 Sri Lanka 75.75 46.92 Azerbaijan 75.75 46.92 Sweden 75.75 46.92 Barbados 75.75 46.92 Switzerland 75.75 46.92 Bahrain 75.75 46.92 Russia 75.75 46.92 Republic of Korea 75.75 46.92 Belgium 73.875 46.92 Belarus 71.68 46.92 Brazil 71.68 46.92 Saint Lucia 71.68 46.92 Armenia 71.68 46.92 Canada 71.68 46.92 Australia 71.68 46.92 United Kingdom 71.68 46.92 Uzbekistan 71.68 46.92 Ukraine 71.68 46.92 Thailand 71.68 46.92 Argentina 71.68 46.92 Saint Kitts and Nevis 71.68 46.92 Bulgaria 71.68 46.92 Belize 71.68 46.92 Netherlands 71.68 46.92 Uruguay 71.68 46.92 Albania 71.68 46.92 Saint Vincent and Grenadines 66.14 46.92 United Arab Emirates 66.14 46.92 Cabo Verde 60.6 46.92 Panama 60.6 46.92 Antigua and Barbuda 60.6 46.92 Paraguay 60.6 46.92 Poland 60.6 46.92 Mauritius 58.65 46.92 Oman 58.65 46.92 New Zealand 58.65 46.92 Philippines 58.65 46.92 Kiribati 46.92 46.92 Chile 41.91 46.92 Germany 41.352000000000004 46.92 Kuwait 40.66 46.92 El Salvador 40.66 46.92 Guyana 40.66 46.92 Luxembourg 40.66 46.92 Latvia 40.66 46.92 Guatemala 40.66 46.92 Malta 40.66 46.92 Georgia 40.56 46.92 Grenada 36.343999999999994 46.92 Lithuania 35.84 46.92 Kyrgyzstan 35.599999999999994 46.92 Macau 35.515 46.92 Ecuador 35.19 46.92 Kazakhstan 31.14 46.92 Greece 30.3 46.92 Colombia 29.3115 46.92 Cuba 29.04 46.92 France 28.831 46.92 Hungary 26.15 46.92 Finland 26.15 46.92 Costa Rica 25.875 46.92 Ireland 24.695 46.92 Czech Republic 24.064999999999998 46.92 Croatia 23.46 46.92 Israel 23.39 46.92 Cyprus 21.93 46.92 Denmark 21.784 46.92 Estonia 17.92 46.92 Maldives 17.808 46.92 Iceland 17.019 46.92 Italy 15.69 46.92 Japan 15.552 46.92 Fiji 14.075 46.92 Jamaica 13.194 46.92 Dominica 10.46 46.92

Turkey has the largest positive difference in average sale value with $146/sale greater than the worldwide average. In contrast, Russia has the largest negative difference with $253/sale less than the world average.

We'll now look at returns of purchased items for each country.

In [9]:
-- CTE for value of sales returned
WITH returned_value AS (
    SELECT country,
        -- Sum of total dollars returned
        SUM(Returned_Amount) AS returned_amount,
        -- Percentage of dollars returned over dollars sold
        -- Sales column doesn't include amount for returned items
        SUM(Returned_Amount) / (SUM(Sales) + SUM(Returned_Amount)) AS pct_value_returned
    From logs
    GROUP BY country),

-- CTE for number of sales returned
returned_sales AS (
    SELECT t1.country, 
        -- counting all rows (number of transactions)
        COUNT(*) AS num_sales,
        -- counting number of order returns 
        COUNT(t2.Returned) AS sales_returned,
        -- calculating percentage of orders that were returned
        1.0 * COUNT(t2.Returned) / COUNT(*) AS pct_sales_returned
    FROM logs as t1
    LEFT JOIN (SELECT duration_seconds, ip, country, Returned
    FROM logs 
    WHERE Returned = 'Yes') AS t2
    ON t1.duration_seconds = t2.duration_seconds
    AND t1.ip = t2.ip
    GROUP BY t1.country)

-- final query with percent of orders returned and percent of sales refunded
SELECT TOP 5 CTE1.country, pct_sales_returned, pct_value_returned 
FROM returned_value AS CTE1
    INNER JOIN returned_sales AS CTE2
    ON CTE1.country = CTE2.country 
ORDER BY pct_sales_returned DESC, pct_value_returned DESC;



Warning: Null value is eliminated by an aggregate or other SET operation.

(5 rows affected)

Total execution time: 00:00:00.220

country pct_sales_returned pct_value_returned Nicaragua 0.212962962962 0.21179128180131437 Oman 0.212962962962 0.15195254761361243 Montenegro 0.190277777777 0.2197341784693778 Puerto Rico 0.184139784946 0.15605512400495508 Poland 0.182870370370 0.14169463345436814

Nicaragua and Oman have the highest percentage of orders returned at 21.3%. This represents a possible opportunity to increase revenue by making the return policy more strict to attempt to lower the number of returned orders. The affect that a stricter return policy would have on customer experience should also be taken into account though.

We will next look at the number of returns for each language. This will help inform us of any language/communication issues thay may be causing a higher rate of purchase returns.

In [10]:
SELECT TOP 5 t1.Languages, 
        -- calculating percentage of orders that were returned
        ROUND((1.0 * COUNT(t2.Returned) / COUNT(*)), 2) AS pct_orders_returned
FROM logs as t1
        LEFT JOIN (SELECT duration_seconds, ip, country, Returned
        FROM logs 
        WHERE Returned = 'Yes') AS t2
        ON t1.duration_seconds = t2.duration_seconds
        AND t1.ip = t2.ip
GROUP BY t1.Languages
ORDER BY pct_orders_returned DESC;

Warning: Null value is eliminated by an aggregate or other SET operation.

(5 rows affected)

Total execution time: 00:00:00.167

Languages pct_orders_returned marathi 0.220000000000 romanian 0.180000000000 kinyarwanda 0.180000000000 icelandic 0.170000000000 polish 0.170000000000

Marathi, a language spoken in India, has the highest percentage for orders returned (4% above the second language). A statistical test should be performed to determine if this percentage is statistically significantly greater than the 2nd highest percentage. If it is, this indicates that higher return rates could be linked to translation issues with the language. This means customers could be misinformed about products, resulting in higher return rates. The impact of language on return rates should be further investigated for the website as a whole to idenitfy if that is a frequent issue.

The membership types associated with online purchases will now be investigated per country. It is important to note that each count of a membership type represents an online purchase but not a unique individual (no unique identifiers are present in dataset). It is likely that premium members make more frequent purchases than normal and non-members. This category therefore investigates purchases per membership type and not individuals per membership type.

In [11]:
-- first CTE for total website purchases per country
WITH totals AS (
    SELECT country, COUNT(*) AS total_purchases
    FROM logs
    GROUP BY country),

-- second CTE for website purchases per membership type per country
memberships AS (
    SELECT country, membership, COUNT(1) AS website_purchase
    FROM logs
    GROUP BY country, membership)

-- joining CTEs and calculating pct membership type for total purchases per country
SELECT memberships.country, membership, website_purchase, 1.0 * website_purchase/total_purchases AS pct_visits
FROM memberships
    LEFT JOIN totals
    ON memberships.country = totals.country
ORDER BY country;

(298 rows affected)

Total execution time: 00:00:00.242

country membership website_purchase pct_visits Albania Not Logged In 17 0.009199134199 Albania Premium 1185 0.641233766233 Albania Normal 646 0.349567099567 Antigua and Barbuda Premium 1527 0.673280423280 Antigua and Barbuda Normal 721 0.317901234567 Antigua and Barbuda Not Logged In 20 0.008818342151 Argentina Not Logged In 24 0.009220130618 Argentina Normal 967 0.371494429504 Argentina Premium 1612 0.619285439877 Armenia Normal 678 0.325023969319 Armenia Not Logged In 20 0.009587727708 Armenia Premium 1388 0.665388302972 Aruba Normal 409 0.347789115646 Aruba Premium 763 0.648809523809 Aruba Not Logged In 4 0.003401360544 Australia Normal 795 0.315476190476 Australia Premium 1684 0.668253968253 Australia Not Logged In 41 0.016269841269 Austria Normal 835 0.312266267763 Austria Not Logged In 12 0.004487658937 Austria Premium 1827 0.683246073298 Azerbaijan Normal 424 0.315476190476 Azerbaijan Not Logged In 6 0.004464285714 Azerbaijan Premium 914 0.680059523809 Bahamas Premium 1354 0.700828157349 Bahamas Not Logged In 10 0.005175983436 Bahamas Normal 568 0.293995859213 Bahrain Not Logged In 11 0.006235827664 Bahrain Normal 495 0.280612244897 Bahrain Premium 1258 0.713151927437 Barbados Premium 1402 0.667937112910 Barbados Not Logged In 10 0.004764173415 Barbados Normal 687 0.327298713673 Belarus Normal 546 0.309699376063 Belarus Premium 1203 0.682359614293 Belarus Not Logged In 14 0.007941009642 Belgium Normal 816 0.313364055299 Belgium Not Logged In 30 0.011520737327 Belgium Premium 1758 0.675115207373 Belize Normal 814 0.346088435374 Belize Premium 1532 0.651360544217 Belize Not Logged In 6 0.002551020408 Bosnia and Herzegovina Not Logged In 2 0.011904761904 Bosnia and Herzegovina Normal 44 0.261904761904 Bosnia and Herzegovina Premium 122 0.726190476190 Brazil Premium 1657 0.636328725038 Brazil Normal 931 0.357526881720 Brazil Not Logged In 16 0.006144393241 Bulgaria Not Logged In 34 0.013492063492 Bulgaria Normal 914 0.362698412698 Bulgaria Premium 1572 0.623809523809 Cabo Verde Premium 57 0.678571428571 Cabo Verde Normal 27 0.321428571428 Canada Premium 1554 0.637931034482 Canada Not Logged In 28 0.011494252873 Canada Normal 854 0.350574712643 Chile Premium 1498 0.575268817204 Chile Not Logged In 249 0.095622119815 Chile Normal 857 0.329109062980 Colombia Not Logged In 579 0.230310262529 Colombia Normal 848 0.337311058074 Colombia Premium 1087 0.432378679395 Costa Rica Premium 947 0.438425925925 Costa Rica Normal 691 0.319907407407 Costa Rica Not Logged In 522 0.241666666666 Croatia Normal 429 0.272900763358 Croatia Not Logged In 378 0.240458015267 Croatia Premium 765 0.486641221374 Cuba Premium 926 0.535879629629 Cuba Normal 452 0.261574074074 Cuba Not Logged In 350 0.202546296296 Cyprus Not Logged In 258 0.241799437675 Cyprus Premium 517 0.484536082474 Cyprus Normal 292 0.273664479850 Czech Republic Not Logged In 513 0.265527950310 Czech Republic Premium 910 0.471014492753 Czech Republic Normal 509 0.263457556935 Denmark Premium 824 0.520202020202 Denmark Not Logged In 390 0.246212121212 Denmark Normal 370 0.233585858585 Dominica Not Logged In 21 0.291666666666 Dominica Normal 25 0.347222222222 Dominica Premium 26 0.361111111111 Ecuador Normal 690 0.309139784946 Ecuador Premium 1152 0.516129032258 Ecuador Not Logged In 390 0.174731182795 El Salvador Not Logged In 299 0.173032407407 El Salvador Premium 870 0.503472222222 El Salvador Normal 559 0.323495370370 Estonia Normal 358 0.236772486772 Estonia Not Logged In 471 0.311507936507 Estonia Premium 683 0.451719576719 Fiji Premium 316 0.398989898989 Fiji Not Logged In 264 0.333333333333 Fiji Normal 212 0.267676767676 Finland Normal 550 0.263409961685 Finland Not Logged In 507 0.242816091954 Finland Premium 1031 0.493773946360 France Premium 1085 0.502314814814 France Normal 669 0.309722222222 France Not Logged In 406 0.187962962962 Georgia Premium 901 0.568813131313 Georgia Normal 460 0.290404040404 Georgia Not Logged In 223 0.140782828282 Germany Not Logged In 265 0.1415598

In [12]:
-- purchases per browser per country
SELECT country, Accessed_From, COUNT(1) AS purchases
FROM logs
GROUP BY country, Accessed_From
ORDER BY country;

(695 rows affected)

Total execution time: 00:00:00.089

country Accessed_From purchases Albania Microsoft Edge 171 Albania Android App 392 Albania Chrome 299 Albania Safari 180 Albania IOS App 233 Albania Others 281 Albania Mozilla Firefox 292 Antigua and Barbuda IOS App 289 Antigua and Barbuda Microsoft Edge 189 Antigua and Barbuda Others 336 Antigua and Barbuda Chrome 401 Antigua and Barbuda Safari 210 Antigua and Barbuda Android App 486 Antigua and Barbuda Mozilla Firefox 357 Argentina Android App 575 Argentina IOS App 336 Argentina Chrome 429 Argentina Others 407 Argentina Safari 240 Argentina Microsoft Edge 216 Argentina Mozilla Firefox 400 Armenia Others 314 Armenia Safari 210 Armenia Microsoft Edge 189 Armenia Mozilla Firefox 332 Armenia IOS App 250 Armenia Chrome 304 Armenia Android App 487 Aruba Chrome 228 Aruba IOS App 155 Aruba Microsoft Edge 99 Aruba Android App 234 Aruba Safari 110 Aruba Mozilla Firefox 164 Aruba Others 186 Australia IOS App 300 Australia Safari 240 Australia Android App 575 Australia Mozilla Firefox 400 Australia Chrome 415 Australia Microsoft Edge 216 Australia Others 374 Austria Chrome 418 Austria Mozilla Firefox 408 Austria Android App 576 Austria IOS App 346 Austria Safari 262 Austria Microsoft Edge 243 Austria Others 421 Azerbaijan Chrome 238 Azerbaijan Safari 128 Azerbaijan IOS App 167 Azerbaijan Others 205 Azerbaijan Microsoft Edge 108 Azerbaijan Android App 298 Azerbaijan Mozilla Firefox 200 Bahamas Microsoft Edge 177 Bahamas Android App 432 Bahamas Mozilla Firefox 306 Bahamas Chrome 296 Bahamas IOS App 243 Bahamas Others 292 Bahamas Safari 186 Bahrain Android App 373 Bahrain Mozilla Firefox 270 Bahrain Others 284 Bahrain Microsoft Edge 147 Bahrain Safari 164 Bahrain IOS App 227 Bahrain Chrome 299 Barbados Microsoft Edge 180 Barbados Others 311 Barbados Safari 211 Barbados Android App 482 Barbados Mozilla Firefox 320 Barbados Chrome 343 Barbados IOS App 252 Belarus Android App 410 Belarus IOS App 220 Belarus Microsoft Edge 153 Belarus Mozilla Firefox 275 Belarus Chrome 256 Belarus Others 279 Belarus Safari 170 Belgium Others 394 Belgium Android App 556 Belgium Mozilla Firefox 396 Belgium Safari 250 Belgium IOS App 325 Belgium Microsoft Edge 225 Belgium Chrome 458 Belize Others 376 Belize Android App 523 Belize Microsoft Edge 207 Belize Mozilla Firefox 359 Belize IOS App 294 Belize Chrome 371 Belize Safari 222 Bosnia and Herzegovina Android App 52 Bosnia and Herzegovina Safari 18 Bosnia and Herzegovina Mozilla Firefox 40 Bosnia and Herzegovina Others 15 Bosnia and Herzegovina IOS App 18 Bosnia and Herzegovina Microsoft Edge 9 Bosnia and Herzegovina Chrome 16 Brazil IOS App 331 Brazil Android App 551 Brazil Safari 249 Brazil Mozilla Firefox 385 Brazil Chrome 460 Brazil Microsoft Edge 225 Brazil Others 403 Bulgaria Android App 576 Bulgaria Safari 240 Bulgaria Microsoft Edge 216 Bulgaria Others 382 Bulgaria Mozilla Firefox 399 Bulgaria Chrome 401 Bulgaria IOS App 306 Cabo Verde Mozilla Firefox 12 Cabo Verde Others 18 Cabo Verde IOS App 28 Cabo Verde Chrome 14 Cabo Verde Microsoft Edge 12 Canada Safari 250 Canada Chrome 377 Canada Mozilla Firefox 372 Canada IOS App 296 Canada Microsoft Edge 213 Canada Android App 552 Canada Others 376 Chile Android App 575 Chile Microsoft Edge 216 Chile Chrome 457 Chile Others 396 Chile Mozilla Firefox 400 Chile Safari 240 Chile IOS App 320 Colombia Mozilla Firefox 386 Colombia IOS App 326 Colombia Safari 240 Colombia Microsoft Edge 225 Colombia Others 399 Colombia Android App 537 Colombia Chrome 401 Costa Rica Microsoft Edge 180 Costa Rica Chrome 357 Costa Rica Safari 210 Costa Rica Mozilla Firefox 324 Costa Rica Android App 483 Costa Rica Others 336 Costa Rica IOS App 270 Croatia Mozilla Firefox 246 Croatia Safari 150 Croatia IOS App 198 Croatia Chrome 255 Croatia Microsoft Edge 138 Croatia Android App 345 Croatia Others 240 Cuba IOS App 198 Cuba Chrome 258 Cuba Safari 180 Cuba Mozilla Firefox 255 Cuba Microsoft Edge 156 Cuba Others 270 Cuba Android App 411 Cyprus Mozilla Firefox 171 Cyprus Microsoft Edge 84 Cypr